resolwe/storage/models.py

```diff
@@ -92,6 +97,201 @@ def get_queryset(self) -> models.QuerySet:
+
+        1. When delay is None return 0. Entries without explicit delay are
+            processed immediately.
+        2. When delay is negative return the number of days between the
+            current date and 1900. There should be no object created before
+            1900 so this means no object will be processed.
+        3. Otherwise return the delay.
+        """
+        if delay is None:
+            return 0
+        assert isinstance(delay, int), "Delay should be integer or None"
+        if delay < 0:
+            return (datetime.now() - datetime(1900, 1, 1)).days
+        else:
+            return delay
+
+    def to_delete(self, connector_name: str) -> models.QuerySet:
+        """Get a queryset of locations that must be deleted.
+
+        :returns: a queryset of FileStorage objects whose StorageLocations objects
+            belonging to the given connector must be deleted.
+        """
+
+        location_settings = copy.deepcopy(STORAGE_CONNECTORS.get(connector_name, {}))
+        rules: dict = location_settings.get("config", {}).get("delete", {})
+
+        if not rules:
+            return FileStorage.objects.none()
+
+        process_overrides = rules.pop("process_type", {})
+        data_slug_overrides = rules.pop("data_slug", {})
+
+        whens = [
+            models.When(
+                data__slug=data_slug,
+                then=timedelta(days=self._preprocess_delay(override_rule.get("delay"))),
+            )
+            for data_slug, override_rule in data_slug_overrides.items()
+        ]
+        whens += [
+            models.When(
+                data__process__type__startswith=process_type,
+                then=timedelta(days=self._preprocess_delay(override_rule.get("delay"))),
+            )
+            for process_type, override_rule in process_overrides.items()
+        ]
+
+        connectors_priority_whens = [
+            models.When(
+                connector_name=connector_name,
+                then=connectors[connector_name].priority,
+            )
+            for connector_name in connectors
+        ]
+
+        min_other_copies = rules.get("min_other_copies", 1)
+        default_delay = self._preprocess_delay(rules.get("delay"))
+
+        return (
+            FileStorage.objects.filter(
+                data__status=Data.STATUS_DONE,
+                created__lt=now()  # type: ignore
+                - models.Case(
+                    *whens,
+                    default=timedelta(days=default_delay),
+                    output_field=models.DurationField(),
+                ),
+            )
+            .annotate(
+                all_locations_count=models.Count(
+                    "storage_locations",
+                    filter=models.Q(
+                        storage_locations__status=StorageLocation.STATUS_DONE
+                    ),
+                    distinct=True,
+                )
+            )
+            .annotate(
```

Comment:
Can you merge all annotations and separate them with commas? Would that give the same result?

Reply:
I see no reason why not. I have merged the annotations into one (and also some filters).